coderd/userauth_test.go

```diff
@@ -1673,27 +1682,186 @@ func TestUserForgotPassword(t *testing.T) {
+		// First try to login before changing our password. We expected this to error
+		// as we haven't change the password yet.
+		_, err := anotherClient.LoginWithPassword(ctx, codersdk.LoginWithPasswordRequest{
+			Email:    anotherUser.Email,
+			Password: newPassword,
+		})
+		require.Error(t, err)
+
+		err = anotherClient.RequestOneTimePasscode(ctx, codersdk.RequestOneTimePasscodeRequest{
 			Email: anotherUser.Email,
 		})
 		require.NoError(t, err)
 
 		require.Equal(t, 2, len(notifyEnq.Sent))
 
 		notif := notifyEnq.Sent[1]
-		require.Equal(t, notifications.TemplateUserRequestedOneTimePasscode, notif.TemplateID)
-		require.Equal(t, anotherUser.ID, notif.UserID)
-		require.Equal(t, 1, len(notif.Targets))
-		require.Equal(t, anotherUser.ID, notif.Targets[0])
+		verifyOneTimePasscodeNotification(t, notif, anotherUser.ID)
 
 		oneTimePasscode := notif.Labels["one_time_passcode"]
 
 		err = anotherClient.ChangePasswordWithOneTimePasscode(ctx, codersdk.ChangePasswordWithOneTimePasscodeRequest{
 			Email:           anotherUser.Email,
 			OneTimePasscode: oneTimePasscode,
+			Password:        newPassword,
+		})
+		require.NoError(t, err)
+
+		// Now that we have changed the password, this should work.
+		_, err = anotherClient.LoginWithPassword(ctx, codersdk.LoginWithPasswordRequest{
+			Email:    anotherUser.Email,
+			Password: newPassword,
+		})
+		require.NoError(t, err)
+
+		// We now need to check that the one-time passcode isn't valid.
+		err = anotherClient.ChangePasswordWithOneTimePasscode(ctx, codersdk.ChangePasswordWithOneTimePasscodeRequest{
+			Email:           anotherUser.Email,
+			OneTimePasscode: oneTimePasscode,
+			Password:        "SomeDifferentSecurePassword!",
+		})
+		require.Error(t, err)
+	})
+
+	t.Run("CannotChangePasswordWithInvalidOneTimePasscode", func(t *testing.T) {
+		t.Parallel()
+
+		notifyEnq := &testutil.FakeNotificationsEnqueuer{}
+
+		client := coderdtest.New(t, &coderdtest.Options{
+			NotificationsEnqueuer: notifyEnq,
+		})
+		user := coderdtest.CreateFirstUser(t, client)
+
+		ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitLong)
+		defer cancel()
+
+		anotherClient, anotherUser := coderdtest.CreateAnotherUser(t, client, user.OrganizationID)
+
+		err := anotherClient.RequestOneTimePasscode(ctx, codersdk.RequestOneTimePasscodeRequest{
+			Email: anotherUser.Email,
+		})
+		require.NoError(t, err)
+
+		require.Equal(t, 2, len(notifyEnq.Sent))
+
+		notif := notifyEnq.Sent[1]
+		verifyOneTimePasscodeNotification(t, notif, anotherUser.ID)
+
+		err = anotherClient.ChangePasswordWithOneTimePasscode(ctx, codersdk.ChangePasswordWithOneTimePasscodeRequest{
+			Email:           anotherUser.Email,
+			OneTimePasscode: uuid.New().String(), // Use a different UUID to the one expected
+			Password:        "SomeNewSecurePassword!",
+		})
+		require.Error(t, err)
+
+		var apiErr *codersdk.Error
+		require.ErrorAs(t, err, &apiErr)
+		require.Equal(t, http.StatusBadRequest, apiErr.StatusCode())
```

Comment:
Ditto; specificity is essential.
Notice how this and the subsequent tests all expect to fail the same way?

Reply:
Yep, have now made them more specific!